packages/yew/src/context.rs

```diff
@@ -64,15 +64,12 @@ impl<T: Clone + PartialEq> ContextProvider<T> {
 
     /// Notify all subscribed consumers and remove dropped consumers from the list.
     fn notify_consumers(&mut self) {
-        let consumers: Vec<Callback<T>> = self
-            .consumers
+        self.consumers
             .borrow()
             .iter()
-            .map(|(_, v)| v.clone())
-            .collect();
```

Comment:
Nope, pretty sure this is intended. `.emit()` would run potentially arbitrary code while the borrow is active. Cloning before emitting drops the borrow in between.

Reply:
It's better if we replace it with [`cloned`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.cloned) which should do the clone lazily but still drop the borrow before running `emit`

Reply:
ok, I have revert this back and add some notes.

Reply:
There are two other places borrow_mut `self.consumers`: `ContextProvider::subscribe_consumer` and `ContextHandle::drop`.
since this function take `&mut self`, so it is guaranteed `subscribe_consumer` will not be called in `.emit`. Is the following code works?
```rust
pub struct ContextProvider<T: Clone + PartialEq + 'static> {
     context: T,
     children: Children,
     consumers: RefCell<Slab<Callback<T>>>,
+    notifying: Cell<bool>,
+    pending_remove: RefCell<Vec<usize>>,
 }

 impl<T: Clone + PartialEq + 'static> Drop for ContextHandle<T> {
     fn drop(&mut self) {
         if let Some(component) = self.provider.get_component() {
-            component.consumers.borrow_mut().remove(self.key);
+            if component.notifying.get() {
+                component.pending_remove.borrow_mut().push(self.key);
+            } else {
+                component.consumers.borrow_mut().remove(self.key);
+            }
         }
     }
 }

fn notify_consumers(&mut self) {
+        self.notifying.set(true);
+        self.consumers.borrow().iter().for_each(|(_, consumer)| {
+                consumer.emit(self.context.clone());
+        });
+        for key in self.pending_remove.borrow_mut().drain(..) {
+            self.consumers.borrow_mut().remove(key);
         }
+        self.notifying.set(false);
```

Reply:
My original code should be work, since `notify_consumers` use `&mut self`, and `consumers` is `RefCell`, not `Rc<RefCell>`, so `.emit` can not  modify `consumers` because that will create another `&self` or `&mut self` which violates Rust's alias rule and is UB. Actually I can even use `self.consumers.get_mut()` instead of `self.consumers.borrow()`.

Reply:
Which is actually a bug, I believe. The `notify_consumers` is run during a mutable borrow of the `ContextProvider` during its `update` lifecycle, hence `&mut self`. At the same time, `Scope::context` dutifully could find the same scope in its `find_parent_scope` traversal upwards, but before calling `subscribe_consumer` *choses to ignore a failure to borrow* when calling `get_component`.

So in summary, if a callback registered on for a `context` itself wants to register during `emit` (again) for the same context, it instead will fail and yew pretends the context is simply not there. Not sure what's the best approach but to me both the current and the "fix" are equally wrong :)

Reply:
I just start read the codebase this weekend, but as I said above, thanks to the Rust type system, the modification is strictly equivalent with the original code, and should not introduce any unexpected behavior. So if it does not work, the current codebase must have a bug somewhere.

Reply:
`notify_consumers` is called from the `changed()` lifecycle event which itself requires `&mut self`.

The issue of not being able to acquire a new handle of current context in the context callback is not solvable without moving the notification of listeners to a scheduler runnable (or by other means that the component state can be released).

Until then, this change makes sense to me.

Reply:
`notify_consumers` is called from the `changed()` lifecycle event which itself requires `&mut self`.

The issue of not being able to acquire a new handle of current context in the context callback is not solvable without moving the notification of listeners to a scheduler runnable (or by other means that the component state can be released).

Until then, this change makes sense to me.

Reply:
> moving the notification of listeners to a scheduler runnable (or by other means that the component state can be released).

Is that possible to do? Sounds like a good fix to make to me.

Side note: you commented this twice, I hid the other comment

Reply:
> Is that possible to do?

It should be possible.

However, breaking the cycle here means that subscribers will see the context update at a later time and we should be careful on when this should happen (which scheduler queue does this belong to?).

Can we do this in a separate pull request so this can get merged?